### PR TITLE
Mic 5086/remove hardcoding columns

### PR DESCRIPTION
**Pull request by albrja**
### Determines and creates table based on inputs and not hard coded values for build_table testing utility function.
- *Category*: Test
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5086

### Changes and notes
-Determines the key and values for parameter and key columns instead of hard coded values

### Testing
All tests pass

### CODE REVIEW

File: src/vivarium/testing_utilities.py [hunk 1]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
     return population
 
 
-def build_table(value, year_start, year_end, columns=("age", "year", "sex", "value")):
-    value_columns = columns[3:]
+def build_table(
+    value,
+    parameter_columns={
+        "age": (0, 125),
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:155]: Are these hard-coded defaults defined somewhere? If not I think they should be (_especially_ b/c 2020 will change basically every year)

File: src/vivarium/testing_utilities.py [hunk 2]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
     return population
 
 
-def build_table(value, year_start, year_end, columns=("age", "year", "sex", "value")):
-    value_columns = columns[3:]
+def build_table(
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:152]: Add type hints

File: src/vivarium/testing_utilities.py [hunk 3]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
...
+):
     if not isinstance(value, list):
         value = [value] * len(value_columns)
 
     if len(value) != len(value_columns):
         raise ValueError("Number of values must match number of value columns")
 
+    # Get product of parameter columns
+    range_parameter_product = {
+        key: list(range(value[0], value[1])) for key, value in parameter_columns.items()
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:169]: This does NOT include the end points (2020 and 125); is that intended

File: src/vivarium/testing_utilities.py [hunk 4]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
...
+    }
+    # Build out dict of items we will need cartesian product of to make dataframe
+    product_dict = dict(range_parameter_product)
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:172]: range_parameter_product is already a dict

File: src/vivarium/testing_utilities.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 """
 
+from itertools import product
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:10]: Why is this file packaged instead of living under the tests module?

File: src/vivarium/testing_utilities.py [hunk 6]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
...
-                r_values = []
-                for v in value:
-                    if v is None:
-                        r_values.append(np.random.random())
-                    elif callable(v):
-                        r_values.append(v(age, sex, year))
-                    else:
-                        r_values.append(v)
-                rows.append([age, age + 1, year, year + 1, sex] + r_values)
+    for p in products:
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:177]: nit: Maybe it's a me thing, but I really dislike single-letter variable names. At least they're not `s`, `n`, or `c` (which mucks up pdb)

File: src/vivarium/testing_utilities.py [hunk 7]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
...
+                r_values.append(np.random.random())
+            elif callable(v):
+                r_values.append(v(p))
+            else:
+                r_values.append(v)
+
+        # Get list of values for rows (index values)
+        key_columns_index_values = p[len(parameter_columns) :]
+        key_columns_index_values = [val for val in key_columns_index_values]
+        # Trasnform parameter column values
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:192]: typo

File: src/vivarium/testing_utilities.py [hunk 8]
```diff
@@ -148,31 +149,62 @@ def _non_crn_build_population(
...
+        parameter_columns_index_values = p[: len(parameter_columns)]
+        # Create intervals for parameter columns. Example year, year+1 for year_start and year_end
+        parameter_columns_index_values = [
+            v for val in parameter_columns_index_values for v in (val, val + 1)
```
Comment (stevebachmeier) [inline src/vivarium/testing_utilities.py:196]: Oh, this is why it's ok to be messing the end points

File: src/vivarium/testing_utilities.py (hunk 3 above)
Comment (stevebachmeier) [reply src/vivarium/testing_utilities.py:169]: nm, it's handled below

File: tests/framework/lookup/test_lookup.py [hunk 9]
```diff
@@ -195,7 +202,13 @@ def test_invalid_data_type_build_table(base_config):
 def test_lookup_table_interpolated_return_types(base_config):
     year_start = base_config.time.start.year
     year_end = base_config.time.end.year
-    data = build_table(lambda age, sex, year: year, year_start, year_end)
+    data = build_table(
+        lambda x: x[0],
+        parameter_columns={
```
Comment (stevebachmeier) [inline tests/framework/lookup/test_lookup.py:207]: I think there should be a test that includes 1 parameter col and more than 2 parameter cols since it's been generalized

**Review by stevebachmeier — APPROVED** [review]

File: tests/framework/artifact/test_manager.py [hunk 10]
```diff
@@ -57,7 +59,12 @@ def test_subset_rows():
 def test_subset_columns():
     values = [0, "red", 100]
     data = build_table(
-        values, 1990, 2010, columns=("age", "year", "sex", "draw", "color", "value")
+        values,
+        parameter_columns={
```
Comment (rmudambi) [inline tests/framework/artifact/test_manager.py:63]: Why did you need to specify the parameter columns here?

**Review by rmudambi — APPROVED** [review]

File: src/vivarium/testing_utilities.py (hunk 1 above)
Comment (albrja) [reply src/vivarium/testing_utilities.py:155]: They are just arbitrary values based relatively on what our normal data looks like. No real data is compared to this these are just values used in this brute force approach to make this dataframe.

File: src/vivarium/testing_utilities.py (hunk 4 above)
Comment (albrja) [reply src/vivarium/testing_utilities.py:172]: This is making a copy so we don't overwrite the original.

File: src/vivarium/testing_utilities.py (hunk 5 above)
Comment (albrja) [reply src/vivarium/testing_utilities.py:10]: I don't understand your question.

File: tests/framework/artifact/test_manager.py (hunk 10 above)
Comment (albrja) [reply tests/framework/artifact/test_manager.py:63]: To change the range of the year column to keep it the same as it was since it is different from the default - 2010 vs 2020

File: src/vivarium/testing_utilities.py (hunk 5 above)
Comment (rmudambi) [reply src/vivarium/testing_utilities.py:10]: Because it is used by VPH. When `vivarium_testing_utils` is more developed, it should probably move there.

File: tests/framework/artifact/test_manager.py (hunk 10 above)
Comment (rmudambi) [reply tests/framework/artifact/test_manager.py:63]: I understand that. I'm saying that the specific endpoints probably don't matter for this test, so you should be able to use the defaults.

**Review by rmudambi — APPROVED** [review]

**Review by patricktnast — APPROVED** [review]
